keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,432 @@
+
+- Interaction with termination grace period
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration (e.g., 5 seconds).
+  3. Set the termination grace period to various values, including:
+     - Equal to the sleep duration
+     - Greater than the sleep duration
+     - Greater than the sleep duration, but reduced to less than the sleep duration at runtime
+  4. For each termination grace period value, delete the pod and observe the time it takes for the container to terminate.
+  5. Verify that the container is terminated after the min(sleep, grace).
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit/e2e tests completed and enabled
+- Documentation is added to demonstrate why this is useful in nginx scenario and how exactly nginx needs to be configured.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Additional e2e tests are completed(if needed)
+
+#### GA
+
+- No negative feedback
+- No bug issues reported
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+The previous PreStop behavior will not be broken. Users can continue to use their hooks as it is.
+To use this enhancement, users need to enable the feature gate, and add sleep action in their prestop hook.
+
+#### Downgrade
+Kube-apiserver will ignore sleep in prestop hook.
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PodLifecycleSleepAction
+  - Components depending on the feature gate: kubelet,kube-apiserver
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node?
+
+###### Does enabling the feature change any default behavior?
+
+No
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+The feature can be disabled in Alpha and Beta versions by restarting kube-apiserver with the feature-gate off. In terms of Stable versions, users can choose to opt-out by not setting the sleep field.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+New pods with sleep action in prestop hook can be created.
+
+###### Are there any tests for feature enablement/disablement?
+
+No
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+The change is opt-in, it doesn't impact already running workloads. But problems with the updated validation logic may cause crashes in the apiserver.
```

Comment:
How would the user determine this is the cause of crashes in the apiserver? Will there be any tests that help prevent this from making it into the release?

Reply:
I think this is misleading (crash), let me update later. If a pod with sleepAction was created and the featue is disabled. And this pod is recreated/updated by a user, the pod's yaml won't pass the validation.
In this case, an error will occur to point out the wrong field, instead of the "crash in the apiserver"

Reply:
It's a hard rule that previously accepted objects must not later fail validation.  When it comes to actual API review, we will ensure that :)

Reply:
> It's a hard rule that previously accepted objects must not later fail validation. When it comes to actual API review, we will ensure that :)

Then I think, we can safely say that this feature will not impact already running workloads?

Reply:
yes